neuland/shared/TCAViewer.h

```diff
@@ -0,0 +1,251 @@
+            }
+
+            if (run == nullptr)
+            {
+                LOG(fatal) << "TCAViewer: No FairRun for" << mBranchName;
+            }
+
+            if (mBranchName.empty())
+            {
+                LOG(fatal) << "TCAViewer: cannot init without branch name specified.";
+            }
+
+            eventHeader = run->GetEventHeader();
+
+            mTCA = (mode == in) ? InputInit(rootMan) : OutputInit(rootMan);
+        }
+
+        [[nodiscard]] auto Get() -> decltype(auto)
+        {
+            if constexpr (mode == in)
+            {
+                return InputGet();
+            }
+            else if constexpr (mode == out)
+            {
+                return OutputGet();
+            }
+        }
+
+        auto GetTCA() -> TClonesArray* { return mTCA; }
+
+        auto CheckCapcitySize() -> bool
+        {
+            auto is_changed = true;
+            auto tempSize = mTCA->Capacity();
+            if (mContCapcitySize)
+            {
+                is_changed = mContCapcitySize != tempSize;
+            }
+            if (is_changed)
+            {
+                mContCapcitySize = tempSize;
+            }
+            return is_changed;
+        }
+
+      private:
+        void DisableGuard() { mSafeGuard = false; }
+        void EnableGuard() { mSafeGuard = true; }
+
+        void Clear()
+        {
+            if (mTCA == nullptr)
+            {
+                LOG(error) << "TCAViewer<out>: Cannot reset non-existed TClonesArray!";
+            }
+            mTCA->Clear("C");
+        }
+
+        auto InputInit(FairRootManager* rootMan) -> TClonesArray*
+        {
+            auto* tTCA = dynamic_cast<TClonesArray*>(rootMan->GetObject(mBranchName.data()));
+            if (tTCA == nullptr)
+            {
+                LOG(fatal) << "TCAViewer<in>: No TClonesArray called " << mBranchName
+                           << "could be obtained from the FairRootManager!";
+            }
+            return tTCA;
+        }
+
+        auto OutputInit(FairRootManager* rootMan) -> TClonesArray*
+        {
+            mTCA_out = std::make_unique<TClonesArray>(DataType::Class_Name());
+            rootMan->Register(mBranchName.data(), "", mTCA_out.get(), kTRUE);
+            return mTCA_out.get();
+        }
+
+        [[nodiscard]] auto InputGet() -> decltype(auto)
+        {
+            if (mTCA == nullptr)
+            {
+                LOG(warn) << "TCAViewer<in>: get data from a non-existed TClonesArray!";
+            }
+            auto is_sucess = CheckCapcitySize() ? SetFirstAddr() : true;
+            return is_sucess ? std::span<DataType* const>(mCont, mTCA->GetEntriesFast()) : std::span<DataType* const>();
+        }
+
+        [[nodiscard]] auto OutputGet() -> TCAOutput_SafeGuard<DataType>
+        {
+            if (mSafeGuard)
+            {
+                LOG(fatal) << "TCAViewer<out>: Get() function can only be called once in the scope. ";
```

Comment:
I think LOG(fatal) aborts, but you might want to make this more explicit.

Do what the editor tells you to do: If you add a exception below the log, and the code analysis says it can't be reached, then you don't need it (or just a comment).

Reply:
You are right. Error handling is my weakest spot in C++. Maybe I should throw an exception instead of using "LOG(fatal)"?